packages/context/src/__tests__/unit/inject.unit.ts

```diff
@@ -93,6 +94,44 @@ describe('function argument injection', () => {
     const meta = describeInjectedArguments(SubTestClass);
     expect(meta.map(m => m.bindingSelector)).to.deepEqual([]);
   });
+
+  // https://github.com/strongloop/loopback-next/issues/2946
+  it('allows custom decorator that returns a new constructor', () => {
+    class HelloController {
+      name: string = 'Leonard';
+    }
+
+    const mixinDecorator = () => <C extends Constructor<object>>(
+      constructor: C,
```

Comment:
I find `constructor` confusing, because it's also a reserved name for class constructors. Let's use `target`, `targetClass` or `targetConstructor` instead.